Samples/src/OracleEmulation/Operations.qs

```diff
@@ -0,0 +1,162 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+
+namespace Microsoft.Quantum.Samples.OracleEmulation
+{
+    open Microsoft.Quantum.Canon;
+    open Microsoft.Quantum.Primitive;
+    open Microsoft.Quantum.Extensions.Emulation;
+
+
+    //////////////////////////////////////////////////////////////////////////
+    // Defining and using simple emulated oracles ////////////////////////////
+    //////////////////////////////////////////////////////////////////////////
+
+    // Declare an `intrinsic` oracle that is implemented in C#. See the
+    // `PermutationOracle.Register` call in the driver for its implementation.
+    operation HalfAnswer(x: Qubit[], y: Qubit[]) : Unit {
+        body intrinsic;
+        adjoint intrinsic;
+    }
+
+    // Define a simple permutation function that is used below to create
+    // another oracle.
+    function DoubleAnswerFunc(x: Int, y: Int) : Int {
+        return 84 ^^^ y;
+    }
+
+    // Measure and print the result.
+    operation MeasureAndDisplay(message: String, register: Qubit[]) : Unit {
+        let answer = MeasureInteger(LittleEndian(register));
+        Message(message + $"{answer}.");
+    }
+
+    // # Summary
+    // Here we demonstrate the use of three simple oracles. Each oracle ignores
+    // the content of the first register and XOR's a constant number into the
+    // second register.
+    //
+    // # Input
+    // ## oracle
+    // A quantum operation that implements an oracle
+    //      $O: \ket{x}\ket{y} \rightarrow \ket{x}\ket{f(x, y)}$.
+    operation RunConstantOracles (oracle: ((Qubit[], Qubit[]) => Unit)) : Unit {
+        Message("Querying the oracles...");
+
+        // Prepare a one-qubit register `x` and an eight-qubit register `y`.
+        // Since all the oracles here ignore x, its length and state do not
+        // matter.
+        using (qubits = Qubit[9]) {
+            let x = qubits[0];
+            let y = qubits[1..8];
+            H(x);
+
+            // Apply an oracle that was passed explicitly by the C# driver.
+            oracle([x], y);
+            MeasureAndDisplay("The answer is ", y);
+
+            // Apply an oracle that was declared as `intrinsic` above and
+            // implemented in the C# driver.
+            HalfAnswer([x], y);
+            MeasureAndDisplay("Half the answer is ", y);
+
+            // Apply an oracle defined in terms of a Q# permutation function.
+            EmulateOracle(DoubleAnswerFunc, [x], y);
```

Comment:
I'd suggest asking @alan-geller for more advice about this line, but my initial impression is that this breaks the operational interpretation of `EmulateOracle` as something that could in principle be written as a set of instructions to a quantum device, as it takes a Q# function and implicitly applies that to the _state_ of its qubit arguments through the use of the C# definition above. If I understand correctly, an operationally consistent interpretation of the Q# program would still be possible, though, if the oracle function was defined at the C# level rather than the Q# level, such that from the perspective of the Q# program, the oracle would be a black box.

Reply:
I don't see a reason why this could not be made to run on a quantum device in principle: `EmulateOracle` is an operation, which could run, say, a reversible synthesis algorithm on the permutation function and then apply the corresponding circuit to its qubit arguments. For all the caller knows, it may act on the state of its qubit arguments by applying a series of gates - just as any other operation does.

Reply:
Should be resolved as per discussion above.